source/server/software_enumerator.cpp

```diff
@@ -0,0 +1,58 @@
+#include "software_enumerator.h"
+
+namespace nidevice_grpc {
+
+SoftwareEnumerator::SoftwareEnumerator(SysCfgLibraryInterface* library)
+    : SysCfgSessionHandler(library)
+{
+}
+
+SoftwareEnumerator::~SoftwareEnumerator()
+{
+}
+
+// Provides a list of installed software on a server under localhost. This internally uses the
+// "NI System Configuration API". If it is not currently installed, it can be downloaded from this page:
+// https://www.ni.com/en-in/support/downloads/drivers/download.system-configuration.html.
+::grpc::Status SoftwareEnumerator::enumerate_software(
+    ::grpc::ServerContext* context,
+    google::protobuf::RepeatedPtrField<SoftwareProperties>* software)
+{
+  NISysCfgStatus status = NISysCfg_OK;
+  NISysCfgSessionHandle session = NULL;
+  NISysCfgEnumSoftwareComponentHandle installedComps = NULL;
+  unsigned int numInstalledComps = 0;
+  char package_id[NISYSCFG_SIMPLE_STRING_LENGTH] = "";
+  char version[NISYSCFG_SIMPLE_STRING_LENGTH] = "";
+  char product_name[NISYSCFG_SIMPLE_STRING_LENGTH] = "";
+
+  try {
+    auto library = get_syscfg_library_interface();
+    if (NISysCfg_Succeeded(status = open_or_get_localhost_syscfg_session(&session))) {
+      if (NISysCfg_Succeeded(status = library->GetInstalledSoftwareComponents(session, NISysCfgIncludeItemsAllVisibleAndHidden, NISysCfgBoolFalse, &installedComps))) {
```

Comment:
Consider: It might be a good idea to have some kind of input/request argument as to whether you want `NISysCfgIncludeItemsAllVisibleAndHidden` or just `NISysCfgIncludeItemsAllVisible`. The latter feels like it would be useful in more places.

Reply:
Sounds good. I added a call for `EnumerateInstalledSoftwareIncludeHiddenPackages` and just used a boolean for now. Feel free to unresolve if something about my implementation seems off.

Reply:
This part is OK, but the way it's exposed in the .proto is what concerns me. I added a comment there to see what others think.